client/pdf_generator.py

```diff
@@ -0,0 +1,186 @@
+			)
+			url = self.application.config['mailer.webserver_url']
+			pdf = self.get_template(url)
+			pdf_file.multiBuild(pdf)
+			self.logger.info('created, check ' + self.config['output_pdf'])
+
+	def signal_send_target(self, _, target):
+		outfile = self.expand_path(self.config['output_pdf'])
+		pdf_file = platypus.SimpleDocTemplate(outfile,
+			pagesize=letter,
+			rightMargin=72,
+			leftMargin=72,
+			topMargin=72,
+			bottomMargin=18
+		)
+		url = self.application.config['mailer.webserver_url'] + '?uid=' + target.uid
+		pdf = self.get_template(url)
+		try:
+			pdf_file.multiBuild(pdf)
+			self.attach_pdf(outfile)
+		except Exception as err:
+			self.logger.error('pdf could not be opened: ' + repr(err))
+		else:
+			self.logger.info('pdf attachement made linking with uid: ' + target.uid)
+
+	def get_template(self, url):
+		logo = self.config['logo']
+		formatted_time = time.ctime()
+		company = self.application.config['mailer.company_name']
+		sender = self.application.config['mailer.source_email_alias']
+
+		story = []
+		click_me = self.config['link_text']
+		link = '<font color=blue><link href="' + str(url) + '">' + click_me + '</link></font>'
+		if self.config['logo']:
+			img = platypus.Image(logo, 2 * inch, inch)
+			story.append(img)
+
+		style_sheet = styles.getSampleStyleSheet()
+		style_sheet.add(styles.ParagraphStyle(name='Justify', alignment=TA_JUSTIFY))
+		ptext = '<font size=10>' + formatted_time + '</font>'
+		story.append(platypus.Spacer(1, 12))
+		story.append(platypus.Paragraph(ptext, style_sheet['Normal']))
+		story.append(platypus.Spacer(1, 12))
+		with open(self.config['template_file'], 'r') as file_h:
+			for line in file_h:
+				story.append(platypus.Paragraph(line, style_sheet['Normal']))
+		story.append(platypus.Spacer(1, 8))
+		story.append(platypus.Paragraph(link, style_sheet['Justify']))
+		story.append(platypus.Spacer(1, 12))
+		ptext = '<font size=10>Sincerely,</font>'
+		story.append(platypus.Paragraph(ptext, style_sheet['Normal']))
+		story.append(platypus.Spacer(1, 12))
+		ptext = '<font size=10>' + sender + '</font>'
+		story.append(platypus.Paragraph(ptext, style_sheet['Normal']))
+		story.append(platypus.Spacer(1, 12))
+		ptext = '<font size=10>' + company + '</font>'
+		story.append(platypus.Paragraph(ptext, style_sheet['Normal']))
+		return story
+
+	def attach_pdf(self, outfile):
+		self.application.config['mailer.attachment_file'] = outfile
+
+	def signal_send_finished(self, _):
+		if not os.path.isfile(self.config['output_pdf']) and os.access(self.config['output_pdf']. os.W_OK):
+			self.logger.error('no pdf file found at: ' + str(self.config['output_pdf']))
+			return
+		self.logger.info('deleting pdf file: ' + str(self.config['output_pdf']))
+		try:
+			os.remove(self.config['output_pdf'])
+		except Exception as err:
+			self.logger.debug('cannot delete created attachment: ' + repr(err))
+		self.application.config['mailer.attachment_file'] = None
+
+	def expand_path(self, outfile, *args, **kwargs):
```

Comment:
This function should be renamed `render_path` or something similar because you're not just expanding the path out but expanding it and rendering it.